source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,204 @@
+namespace {
+
+bool isCacheableRequest(Http::HeaderMap& headers) {
+  const Http::HeaderEntry* method = headers.Method();
+  const Http::HeaderEntry* scheme = headers.Scheme();
+  const Http::HeaderValues& header_values = Http::Headers::get();
+  // TODO(toddmgreer) Also serve HEAD requests from cache.
+  // TODO(toddmgreer) Check all the other cache-related headers.
+  return method && scheme && headers.Path() && headers.Host() &&
+         (method->value() == Http::Headers::get().MethodValues.Get) &&
+         (scheme->value() == header_values.SchemeValues.Http ||
+          scheme->value() == header_values.SchemeValues.Https);
+}
+
+bool isCacheableResponse(Http::HeaderMap& headers) {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  // TODO(toddmgreer) fully check for cacheability. See for example
+  // https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/http/caching_headers.h.
+  if (cache_control) {
+    return !StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                      Http::Headers::get().CacheControlValues.Private);
+  }
+  return false;
+}
+
+HttpCache& getCache(const envoy::config::filter::http::cache::v2alpha::Cache& config) {
+  HttpCacheFactory* factory =
+      Registry::FactoryRegistry<HttpCacheFactory>::getFactory(config.name());
+  if (!factory) {
+    throw ProtoValidationException(
+        fmt::format("Didn't find a registered HttpCacheFactory for '{}'", config.name()), config);
+  }
+  return factory->getCache();
+}
+} // namespace
+
+CacheFilter::CacheFilter(const envoy::config::filter::http::cache::v2alpha::Cache& config,
+                         const std::string&, Stats::Scope&, TimeSource& time_source)
+    : time_source_(time_source), cache_(getCache(config)) {}
+
+void CacheFilter::onDestroy() {
+  lookup_ = nullptr;
+  insert_ = nullptr;
+}
+
+Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  if (!isCacheableRequest(headers)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ASSERT(decoder_callbacks_);
+  lookup_ = cache_.makeLookupContext(LookupRequest(headers, time_source_.systemTime()));
+  ASSERT(lookup_);
+
+  CacheFilterSharedPtr self = shared_from_this();
+  lookup_->getHeaders([self](LookupResult&& result) { onHeadersAsync(self, std::move(result)); });
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::HeaderMap& headers, bool end_stream) {
+  if (lookup_ && isCacheableResponse(headers)) {
+    insert_ = cache_.makeInsertContext(std::move(lookup_));
+    insert_->insertHeaders(headers, end_stream);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (insert_) {
+    // TODO(toddmgreer) Wait for the cache if necessary.
+    insert_->insertBody(
+        data, [](bool) {}, end_stream);
```

Comment:
does 'format' not put this on a single line?

Reply:
At least on my system, format insists on breaking the line.